examples/petsc/bps.c

```diff
@@ -397,39 +457,94 @@ int main(int argc, char **argv) {
     CeedVectorSetArray(xcoord, CEED_MEM_HOST, CEED_OWN_POINTER, xloc);
   }
 
-  // Create the Q-function that builds the diff operator (i.e. computes its
+  // Create the Q-function that builds the operator (i.e. computes its
   // quadrature data) and set its context data.
-  CeedQFunctionCreateInterior(ceed, 1,
-                              Setup, __FILE__ ":Setup", &qf_setup);
+  switch (bpChoice) {
+    case CEED_BP1:
+      CeedQFunctionCreateInterior(ceed, 1,
+                                  SetupMass, __FILE__ ":SetupMass", &qf_setup);
+      break;
+    case CEED_BP2:
+      CeedQFunctionCreateInterior(ceed, 1,
+                                  SetupMass3, __FILE__ ":SetupMass3", &qf_setup);
+      break;
+    case CEED_BP3:
+    case CEED_BP5:
+      CeedQFunctionCreateInterior(ceed, 1,
+                                  SetupDiff, __FILE__ ":SetupDiff", &qf_setup);
+      break;
+    case CEED_BP4:
+    case CEED_BP6:
+      CeedQFunctionCreateInterior(ceed, 1,
+                                  SetupDiff3, __FILE__ ":SetupDiff3", &qf_setup);
+  }
   CeedQFunctionAddInput(qf_setup, "x", 3, CEED_EVAL_INTERP);
   CeedQFunctionAddInput(qf_setup, "dx", 3, CEED_EVAL_GRAD);
   CeedQFunctionAddInput(qf_setup, "weight", 1, CEED_EVAL_WEIGHT);
-  CeedQFunctionAddOutput(qf_setup, "rho", 6, CEED_EVAL_NONE);
-  CeedQFunctionAddOutput(qf_setup, "true_soln", 1, CEED_EVAL_NONE);
-  CeedQFunctionAddOutput(qf_setup, "rhs", 1, CEED_EVAL_INTERP);
-
-  // Create the Q-function that defines the action of the diff operator.
-  CeedQFunctionCreateInterior(ceed, 1,
-                              Diff, __FILE__ ":Diff", &qf_diff);
-  CeedQFunctionAddInput(qf_diff, "u", 1, CEED_EVAL_GRAD);
-  CeedQFunctionAddInput(qf_diff, "rho", 6, CEED_EVAL_NONE);
-  CeedQFunctionAddOutput(qf_diff, "v", 1, CEED_EVAL_GRAD);
+  if (bpChoice == CEED_BP1 || bpChoice == CEED_BP2)
+    CeedQFunctionAddOutput(qf_setup, "rho", 1, CEED_EVAL_NONE);
+  else
+    CeedQFunctionAddOutput(qf_setup, "rho", 6, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "true_soln", vscale, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "rhs", vscale, CEED_EVAL_INTERP);
+
+  // Set up PDE operator
+  switch (bpChoice) {
```

Comment:
Rather than the sequence of switch statements, it might be cleaner to have an array of structs or even just helper functions for each bp.

Reply:
I'll think on how to rearrange this, but I agree that it can be better.

Reply:
Much cleaner now